In [8]:
from ultralytics import YOLO
import os
import subprocess
from ultralytics import YOLO
import torch

os.environ['CUDA_VISIBLE_DEVICES'] = '0' 

Ниже код, где я обучал модель на 2000 размеченных картинках, сейчас он закомментирован, потому что в самом конце представленно итоговое обучение (уже на 4000 картинках)

In [12]:
# print(f"Torch version: {torch.__version__}")
# print(f"CUDA available: {torch.cuda.is_available()}")

# model = YOLO("yolov8m.pt")
# t_bank_path = "tbank.yaml"

# model.train(
#     data=t_bank_path,
#     epochs=30,
#     imgsz=640,
#     device='cuda'
# )

## Модель, обученная на 2000 размеченных картинках

In [ ]:
model_path = "C:/Users/roker/VSCode_projects/T_Bank_CV_API/T_Bank_CV_API/runs/detect/train4/weights/best.pt"
model = YOLO(model_path)
t_bank_path = "tbank.yaml"
print(f"Загружена модель: {model_path}")


metrics = model.val(data=t_bank_path, iou=0.5)
print(f"mAP50: {metrics.box.map50:.4f}")
print(f"mAP50-95: {metrics.box.map:.4f}")
print(f"Precision: {metrics.box.mp:.4f}")
print(f"Recall: {metrics.box.mr:.4f}")

Загружена модель: C:/Users/roker/VSCode_projects/T_Bank_CV_API/T_Bank_CV_API/runs/detect/train4/weights/best.pt
Ultralytics 8.3.197  Python-3.12.4 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
Model summary (fused): 92 layers, 25,840,339 parameters, 0 gradients, 78.7 GFLOPs
val: Fast image access  (ping: 0.00.0 ms, read: 637.8687.2 MB/s, size: 32.2 KB)
val: Scanning C:\Users\roker\VSCode_projects\T_Bank_CV_API\T_Bank_CV_API\data\splits\labels\val.cache... 141 images, 122 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 141/141  0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 9/9 2.7it/s 3.4s0.3ss
                   all        141         25      0.998       0.96      0.985      0.799
Speed: 0.2ms preprocess, 13.1ms inference, 0.0ms loss, 0.5ms postprocess per image
Results saved to C:\Users\roker\VSCode_projects\T_Bank_CV_API\T_Bank_CV_API\runs\detect\val5
mAP50: 0.9853
mAP50-95: 0.7987
Precision: 0.9981

- mAP50: 0.9853
- mAP50-95: 0.7987
- Precision: 0.9981
- Recall: 0.9600

Мы получили хорошие результаты на 2000 размеченных картинках. Используя файл creating_labels.py обогатим наши данные: пусть наша обученная модель разметит картинки, я проверю их, а потом подгружу в train и val

У меня есть силы проверить еще 2000 изображений, их я только и добавлю. Добавил бы больше то и результат получил бы лучше, но он и так неплохой! 

In [6]:
metrics = model.val(data=t_bank_path,split='test', iou=0.5)
print(f"mAP50: {metrics.box.map50:.4f}")
print(f"mAP50-95: {metrics.box.map:.4f}")
print(f"Precision: {metrics.box.mp:.4f}")
print(f"Recall: {metrics.box.mr:.4f}")

Ultralytics 8.3.197  Python-3.12.4 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
val: Fast image access  (ping: 0.00.0 ms, read: 210.6231.8 MB/s, size: 21.2 KB)
val: Scanning C:\Users\roker\VSCode_projects\T_Bank_CV_API\T_Bank_CV_API\data\splits\labels\test... 1904 images, 1651 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1904/1904 4.3Kit/s 0.4s0.0s
val: New cache created: C:\Users\roker\VSCode_projects\T_Bank_CV_API\T_Bank_CV_API\data\splits\labels\test.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 119/119 4.7it/s 25.4s0.2s
                   all       1904        296      0.994      0.878       0.95      0.889
Speed: 0.2ms preprocess, 12.4ms inference, 0.0ms loss, 0.3ms postprocess per image
Results saved to C:\Users\roker\VSCode_projects\T_Bank_CV_API\T_Bank_CV_API\runs\detect\val4
mAP50: 0.9504
mAP50-95: 0.8891
Precision: 0.9940
Recall: 0.8784


Сначала попробуем протестировать уже обученную модель (в test добавил 2000 проверенных изображений)

- mAP50: 0.9504
- mAP50-95: 0.8891
- Precision: 0.9940
- Recall: 0.8784

В целом, также хорошо. А теперь закинем эти 2000 изображений в train. Обучим, это и будет наша финальная модель. Всего для обучения использовалось 4000 размеченных изображений (2000 мною и 2000 моделью)

## Модель, обученная на 4000 изображений

In [10]:
print(f"Torch version: {torch.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")

model = YOLO("yolov8m.pt")
t_bank_path = "tbank.yaml"

model.train(
    data=t_bank_path,
    epochs=70,
    imgsz=640,
    device='cuda'
)

Torch version: 2.5.1+cu121
CUDA available: True
New https://pypi.org/project/ultralytics/8.3.198 available  Update with 'pip install -U ultralytics'
Ultralytics 8.3.197  Python-3.12.4 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4060 Laptop GPU, 8188MiB)
engine\trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=tbank.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=70, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8m.pt, momentum=0.937, mosaic=1.0, multi_scale

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x000001EB10DEF6E0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.0480

## Итог

Только что обучили модельку. Нет сил размечать еще картинки, чтобы проверить какая модель лучше (на 4000 или на 2000), скорее всего та, что на 4000. Теперь это наша основная модель, её метрики примерно такие:
- mAP50: ~ 0.95
- mAP50-95: ~ 0.89

(сужу по результатом на тестовой выборке для прошлой, менее крутой модели)
